From 0245d66f72e0645a448e2d03a7b957bcc18d919f Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Wed, 26 Nov 2014 21:22:03 +0000
Subject: [PATCH] [CONFIGURATION-564] Properties without a value and separator
 are now supported.

The regular expression was slightly modified to support property
declarations that do not have a separator. Such keys are now added to the
configuration without a value.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@1641943 13f79535-47bb-0310-9956-ffa450edef68
---
 .../PropertiesConfiguration.java              |  4 +--
 .../TestPropertiesConfiguration.java          | 32 +++++++++++++++----
 src/test/resources/test.properties            |  1 +
 3 files changed, 29 insertions(+), 8 deletions(-)

diff --git a/src/main/java/org/apache/commons/configuration2/PropertiesConfiguration.java b/src/main/java/org/apache/commons/configuration2/PropertiesConfiguration.java
index e4875f62c2..2052cf6019 100644
--- a/src/main/java/org/apache/commons/configuration2/PropertiesConfiguration.java
+++ b/src/main/java/org/apache/commons/configuration2/PropertiesConfiguration.java
@@ -628,7 +628,7 @@ public static class PropertiesReader extends LineNumberReader
         private static final Pattern PROPERTY_PATTERN = Pattern
                 .compile("(([\\S&&[^\\\\" + new String(SEPARATORS)
                         + "]]|\\\\.)*)(\\s*(\\s+|[" + new String(SEPARATORS)
-                        + "])\\s*)(.*)");
+                        + "])\\s*)?(.*)");
 
         /** Constant for the index of the group for the key. */
         private static final int IDX_KEY = 1;
@@ -1185,7 +1185,7 @@ public void writeln(String s) throws IOException
         protected String fetchSeparator(String key, Object value)
         {
             return (getGlobalSeparator() != null) ? getGlobalSeparator()
-                    : getCurrentSeparator();
+                    : StringUtils.defaultString(getCurrentSeparator());
         }
     } // class PropertiesWriter
 
diff --git a/src/test/java/org/apache/commons/configuration2/TestPropertiesConfiguration.java b/src/test/java/org/apache/commons/configuration2/TestPropertiesConfiguration.java
index 3e7af5bd94..f68a5aa967 100644
--- a/src/test/java/org/apache/commons/configuration2/TestPropertiesConfiguration.java
+++ b/src/test/java/org/apache/commons/configuration2/TestPropertiesConfiguration.java
@@ -142,15 +142,35 @@ public void testAppend() throws Exception
     }
 
     /**
-     * Tests that empty properties are treated as the empty string
-     * (rather than as null).
+     * Checks for a property without a value.
+     * 
+     * @param key the key to be checked
+     */
+    private void checkEmpty(String key)
+    {
+        String empty = conf.getString(key);
+        assertNotNull("Property not found: " + key, empty);
+        assertEquals("Wrong value for property " + key, "", empty);
+    }
+
+    /**
+     * Tests that empty properties are treated as the empty string (rather than
+     * as null).
+     */
+    @Test
+    public void testEmpty()
+    {
+        checkEmpty("test.empty");
+    }
+
+    /**
+     * Tests that properties are detected that do not have a separator and a
+     * value.
      */
     @Test
-    public void testEmpty() throws Exception
+    public void testEmptyNoSeparator()
     {
-        String empty = conf.getString("test.empty");
-        assertNotNull(empty);
-        assertEquals("", empty);
+        checkEmpty("test.empty2");
     }
 
     /**
diff --git a/src/test/resources/test.properties b/src/test/resources/test.properties
index 704d4ba0ad..74c80c5d04 100644
--- a/src/test/resources/test.properties
+++ b/src/test/resources/test.properties
@@ -32,6 +32,7 @@ test.unescape.list-separator = This string contains \, an escaped list separator
 test.equals = value=one
 
 test.empty =
+test.empty2
 
 test.mixed.array = a
 test.mixed.array = b, c, d
